pilot/pkg/xds/endpoint_builder.go

```diff
@@ -315,9 +323,11 @@ func (b *EndpointBuilder) buildLocalityLbEndpointsFromShards(
 				}
 				localityEpMap[ep.Locality.Label] = locLbEps
 			}
-			if ep.EnvoyEndpoint == nil {
-				ep.EnvoyEndpoint = buildEnvoyLbEndpoint(ep)
+			tlsMode := ep.TLSMode
+			if b.mtlsChecker != nil && b.mtlsChecker.mtlsDisabledByPeerAuthentication(ep) {
+				tlsMode = ""
 			}
+			ep.EnvoyEndpoint = buildEnvoyLbEndpoint(ep, tlsMode)
```

Comment:
should check the impact of this change (ie running benchmark, which I triggered)

Reply:
sure. anything else you would suggest besides the benchmark results?

Reply:
```
EndpointGeneration/1/100-8                           376kB ± 0%         536kB ± 0%  +42.52%  (p=0.016 n=5+4)
EndpointGeneration/10/10-8                           246kB ± 0%         364kB ± 0%  +48.44%  (p=0.008 n=5+5)
EndpointGeneration/100/10-8                         2.33MB ± 0%        3.47MB ± 0%  +48.80%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-8                         2.34MB ± 0%        3.47MB ± 0%  +48.53%  (p=0.008 n=5+5)
```

Not great... do you think its feasible to rebuild only when tlsMode changes (which is very rare)?

Reply:
We definitely can do that(only rebuild when tls Mode changes), but that might not be that rare since now, `tlsMode` value can change caused by a peer authn config changed, right? See my previous comments.

Current code is building policy checker for each endpoint, which underneaths creates the PolicyApplier. I think reuse a single one for a single PushContext would reduce the memory significantly. From config perpective i think this would work as well since the policy applier is not endpoint specific.

Reply:
I think it is rare - we push endpoints FAR more often then peer auth changes. Likely many many orders of magnitude more often.

For PolicyApplier - doesn't it take `ep.Labels` into account?

Reply:
> I think it is rare in relationship to how often we hit this code path. We push endpoints FAR more often then peer auth changes. Likely many many orders of magnitude more often.

I see. you mean like when a single pod comes up we hit this code(rebuild the entire endpoint list for a given cluster maybe) which is far more often than the peer authn changes. In that case, if we only add these cost(cpu/memory) when peer authn changes, that could be acceptable.

true, the applier indeed take the labels into account, though we can build another map from labels to the applier instance to de-dup among the same set of endpoints.

Reply:
did some profilng againt the benchmark test. it appears that [1] the delta inc for the memory indeed comes the `buildEnvoyEndpoint`. conditionally construct the endpoint solves the problem. benchmark numbers should go back to the same level before.

![image](https://user-images.githubusercontent.com/1661831/126400922-7d5fa615-0969-4d5b-b5cf-a9664b6f75e4.png)

Reply:
now fixed. see benchmark test results.

above comment is not enough. we still need to ensure when tlsMode == "", do not set filter metadata, just leave empty values. Because later on EndpointByNetworkFilters would use `proto.Clone`, increase the consumption as well.